### PR TITLE
Add grab image action and camera availability diagnostics

### PR DESCRIPTION
**Pull request by brinij**
This PR will add capabilities:

- grab depth and color image action

- publish `camera_availabilty` diagnostics

- reconnect with the camera if one gets disconnected

- use camera's serial number, provided as device_id, as a distinguishing method when multiple cameras are present

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -184,7 +188,7 @@ install(FILES 56-orbbec-usb.rules
         DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
 )
 
-install(DIRECTORY scripts
+install(DIRECTORY scripts launch
```
Comment (nlamprian) [inline CMakeLists.txt]: Reorder alphabetically

File: include/astra_camera/astra_driver.h [hunk 2]
```diff
@@ -69,13 +69,81 @@
...
 namespace astra_wrapper
 {
 
+typedef struct
+{
+  int exposure;
+  int gain;
+  bool auto_exposure;
+} CameraParameters;
+/**
```
Comment (nlamprian) [inline include/astra_camera/astra_driver.h:88]: Leave space

File: include/astra_camera/astra_driver.h [hunk 3]
```diff
@@ -69,13 +69,81 @@
...
+#include <actionlib/server/simple_action_server.h>
+#include <camera_control_msgs/GrabImagesAction.h>
+
+#include <diagnostic_updater/diagnostic_updater.h>
+#include <diagnostic_updater/publisher.h>
+
 namespace astra_wrapper
 {
 
+typedef struct
```
Comment (nlamprian) [inline include/astra_camera/astra_driver.h]: You could define this the c++ way: `struct CameraParameteres {};`

File: include/astra_camera/astra_driver.h [hunk 4]
```diff
@@ -101,9 +169,10 @@ class AstraDriver
   void initDevice();
 
   void advertiseROSTopics();
+  void closeROSTopics();
 
-  void imageConnectCb();
-  void depthConnectCb();
+  void imageConnectCb(bool grab_image_client_ = false);
+  void depthConnectCb(bool grab_image_client_ = false);
```
Comment (nlamprian) [inline include/astra_camera/astra_driver.h]: Remove trailing underscore (non-member variable), and maybe use plural in image.

File: include/astra_camera/astra_driver.h [hunk 5]
```diff
@@ -69,13 +69,81 @@
...
+};
+
 class AstraDriver
 {
+  typedef actionlib::SimpleActionServer<camera_control_msgs::GrabImagesAction> ImageActionServer;
+  typedef boost::shared_ptr<ImageActionServer> ImageActionServerPtr;
+
 public:
   AstraDriver(ros::NodeHandle& n, ros::NodeHandle& pnh) ;
+  diagnostic_updater::Updater diagnostics_updater_;
```
Comment (nlamprian) [inline include/astra_camera/astra_driver.h]: Make private (move below with the other members)

File: include/astra_camera/astra_driver.h [hunk 6]
```diff
@@ -186,6 +255,8 @@ class AstraDriver
 
   std::map<int, AstraVideoMode> video_modes_lookup_;
 
+  std::string serial_number;
```
Comment (nlamprian) [inline include/astra_camera/astra_driver.h]: Add trailing underscore

File: include/astra_camera/astra_driver.h [hunk 7]
```diff
@@ -217,6 +288,27 @@ class AstraDriver
...
+
+  CameraParameters getCameraParameters() const;
+  void setCameraParameters(const CameraParameters &parameters);
+
+  ImageActionServerPtr color_action_server_;
+  ImageActionServerPtr depth_action_server_;
+  ActionImageSyncer color_syncer_;
+  ActionImageSyncer depth_syncer_;
+
+  void diagnostics_timer_callback_(const ros::TimerEvent&);
```
Comment (nlamprian) [inline include/astra_camera/astra_driver.h]: Write name in camel case (for create_diagnostics as well)

File: include/astra_camera/astra_driver.h [hunk 8]
```diff
@@ -217,6 +288,27 @@ class AstraDriver
...
+  ros::Timer diagnostics_trigger_;
```
Comment (nlamprian) [inline include/astra_camera/astra_driver.h]: Maybe stick to diagnostics_timer_?

File: src/astra_driver.cpp [hunk 9]
```diff
@@ -44,6 +44,8 @@
 #include <boost/date_time/posix_time/posix_time.hpp>
 #include <boost/thread/thread.hpp>
 
+using diagnostic_msgs::DiagnosticStatus;
```
Comment (nlamprian) [inline src/astra_driver.cpp:47]: You are not taking advantage of this. Either remove the `using` and the `diagnostic_msgs` instances in the code.

File: src/astra_driver.cpp [hunk 10]
```diff
@@ -152,12 +153,129 @@ AstraDriver::AstraDriver(ros::NodeHandle& n, ros::NodeHandle& pnh) :
...
+  }
+  else
+  {
+    stat.summary(diagnostic_msgs::DiagnosticStatus::OK, "Device is connected");
+  }
+
+  std::string uri = "";
+
+  // get uri for serial number (changes after losing connection)
+  // todo use std::find
```
Comment (nlamprian) [inline src/astra_driver.cpp]: For me, this looks fine. I would remove the todo

File: src/astra_driver.cpp [hunk 11]
```diff
@@ -152,12 +153,129 @@ AstraDriver::AstraDriver(ros::NodeHandle& n, ros::NodeHandle& pnh) :
...
+  {
+    // first connection
+    device_uri_ = uri;
+    return;
+  }
+  else
+  {
+    // camera was reconnected and gets an incremented URI
+    // todo fix compare
+    if (device_uri_.compare(uri) != 0)
```
Comment (nlamprian) [inline src/astra_driver.cpp]: device_uri_ != uri

File: src/astra_driver.cpp [hunk 12]
```diff
@@ -152,12 +153,129 @@ AstraDriver::AstraDriver(ros::NodeHandle& n, ros::NodeHandle& pnh) :
   ROS_DEBUG("Dynamic reconfigure configuration received.");
 
   advertiseROSTopics();
+
+  diagnostics_updater_.verbose_ = false;
+  diagnostics_updater_.setHardwareID("astra_" + serial_number);
+  diagnostics_updater_.add("camera_availability", this, &AstraDriver::create_diagnostics);
```
Comment (nlamprian) [inline src/astra_driver.cpp]: Maybe a more suitable name would be getAvailabilityDiagnostic

File: src/astra_driver.cpp [hunk 13]
```diff
@@ -152,12 +153,129 @@ AstraDriver::AstraDriver(ros::NodeHandle& n, ros::NodeHandle& pnh) :
...
+    }
+  } 
+}
+
+void AstraDriver::diagnostics_timer_callback_(const ros::TimerEvent &)
+{
+  diagnostics_updater_.update();
+}
+
+void AstraDriver::closeROSTopics()
```
Comment (nlamprian) [inline src/astra_driver.cpp]: Maybe use a more appropriate verb? terminate/unsubscribe/shutdown/...

File: src/astra_driver.cpp [hunk 14]
```diff
@@ -464,6 +583,192 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+    result.reached_brightness_values.push_back(NAN);
+    result.reached_gamma_values.push_back(NAN);
+
+    // Send feedback
+    ++feedback.curr_nr_images_taken;
+    color_action_server_->publishFeedback(feedback);
+  }
+
+  // Finalize result
+  // Todo use empty()
```
Comment (nlamprian) [inline src/astra_driver.cpp]: Apply todo

File: src/astra_driver.cpp [hunk 15]
```diff
@@ -464,6 +583,192 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+  if (result.images.size() != 0)
+  {
+    sensor_msgs::Image image = result.images.front();
```
Comment (nlamprian) [inline src/astra_driver.cpp]: const sensor_msgs::Image&

File: src/astra_driver.cpp [hunk 16]
```diff
@@ -464,6 +583,192 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+    result.images.push_back(*depth_syncer_.image);
+
+    // Set parameters
+    result.reached_exposure_times.push_back(NAN);
+    result.reached_gain_values.push_back(NAN);
+    result.reached_brightness_values.push_back(NAN);
+    result.reached_gamma_values.push_back(NAN);
+
+    // Send feedback
+    camera_control_msgs::GrabImagesFeedback feedback;
```
Comment (nlamprian) [inline src/astra_driver.cpp:736]: Use as in `colorGrabImagesCallback` and increment the variable

File: src/astra_driver.cpp [hunk 17]
```diff
@@ -464,6 +583,192 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+    feedback.curr_nr_images_taken = 1;
+    depth_action_server_->publishFeedback(feedback);
+  }
+
+  // Finalize result
+  // Todo use empty
```
Comment (nlamprian) [inline src/astra_driver.cpp]: Apply todo

File: src/astra_driver.cpp [hunk 18]
```diff
@@ -464,6 +579,190 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+    result.reached_brightness_values.push_back(NAN);
+    result.reached_gamma_values.push_back(NAN);
+
+    // Send feedback
+    ++feedback.curr_nr_images_taken;
+    color_action_server_->publishFeedback(feedback);
+  }
+
+  // Finalize result
+  if (result.images.empty())
```
Comment (nlamprian) [inline src/astra_driver.cpp]: not empty (below as well)

Comment (brinij) [reply src/astra_driver.cpp]: thanks!

**Review by nlamprian — APPROVED** [review]
lgtm

File: CMakeLists.txt [hunk 19]
```diff
@@ -132,6 +132,10 @@ target_link_libraries(astra_list_devices astra_wrapper)
 
 add_executable(astra_test_wrapper test/test_wrapper.cpp )
 target_link_libraries(astra_test_wrapper astra_wrapper ${Boost_LIBRARIES})
+
+add_executable(action_image_client test/action_image_client.cpp)
```
Comment (g-gemignani) [inline CMakeLists.txt:136]: why did you add the image client to the test folder?

File: src/astra_driver.cpp (hunk 9 above)
Comment (g-gemignani) [reply src/astra_driver.cpp:47]: still to be addressed?

Comment (brinij) [reply src/astra_driver.cpp:47]: No, it is changed. Just not in this line of code. I left `using` and removed `diganostic_msgs::`

File: CMakeLists.txt [hunk 20]
```diff
@@ -68,7 +68,7 @@ generate_messages(
 catkin_package(
   INCLUDE_DIRS include
   LIBRARIES astra_wrapper
-  CATKIN_DEPENDS camera_info_manager dynamic_reconfigure image_transport nodelet sensor_msgs roscpp message_runtime
+  CATKIN_DEPENDS camera_info_manager actionlib camera_control_msgs dynamic_reconfigure image_transport nodelet sensor_msgs roscpp message_runtime
```
Comment (dorezyuk) [inline CMakeLists.txt]: diagnostic_updater is missing

File: CMakeLists.txt (hunk 19 above)
Comment (brinij) [reply CMakeLists.txt:136]: Because that is the client that publishes images to debug_images topic in order to test rgb/grab_images action. Where would you put it?

Comment (g-gemignani) [reply CMakeLists.txt:136]: Where you usually add node executables... In my brain in test you add only tests... but maybe it's just my brain :)

**Review by g-gemignani — APPROVED** [review]

File: src/astra_driver.cpp [hunk 21]
```diff
@@ -152,12 +153,127 @@ AstraDriver::AstraDriver(ros::NodeHandle& n, ros::NodeHandle& pnh) :
...
+
+    terminateROSTopics();
+    initDevice(); // blocking
+  }
+  else
+  {
+    stat.summary(DiagnosticStatus::OK, "Device is connected");
+  }
+
+  std::string uri = "";
```
Comment (dorezyuk) [inline src/astra_driver.cpp]: just uri will construct an empty string for you
https://en.cppreference.com/w/cpp/string/basic_string/basic_string

File: src/astra_driver.cpp [hunk 22]
```diff
@@ -152,12 +153,127 @@ AstraDriver::AstraDriver(ros::NodeHandle& n, ros::NodeHandle& pnh) :
...
+
+  // get uri for serial number (changes after losing connection)
+  std::vector<AstraDeviceInfo> infos = *device_manager_->getConnectedDeviceInfos();
```
Comment (dorezyuk) [inline src/astra_driver.cpp]: this will get you a copy of the connected devices. why not taking the ref? can this variable change underneath without noticing?

File: src/astra_driver.cpp [hunk 23]
```diff
@@ -152,12 +153,127 @@ AstraDriver::AstraDriver(ros::NodeHandle& n, ros::NodeHandle& pnh) :
...
+  }
+
+  // we have a uri for a camera with the right serial number
+  stat.summary(DiagnosticStatus::OK, "Device is connected");
+
+  if (device_uri_.empty())
+  {
+    // first connection
+    device_uri_ = uri;
+    return;
```
Comment (dorezyuk) [inline src/astra_driver.cpp:226]: returns are redundant (here and view lines below)

File: src/astra_driver.cpp [hunk 24]
```diff
@@ -464,6 +579,190 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
   }
 }
 
+bool AstraDriver::validateGrabImagesGoal(const camera_control_msgs::GrabImagesGoalConstPtr &goal)
+{
+  if (goal->exposure_given and goal->gain_given)
```
Comment (dorezyuk) [inline src/astra_driver.cpp]: please go for consistency and use && || and ! throughout the lib

File: src/astra_driver.cpp [hunk 25]
```diff
@@ -464,6 +579,190 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+    ROS_WARN_ONCE("Received goal with gamma or brightness given. These values "
+                  "are not supported and will be ignored!");
+  }
+
+  imageConnectCb(true); // Connect (in case there are no subscribers)
+  CameraParameters old_parameters = getCameraParameters(); // Backup parameters
+
+  camera_control_msgs::GrabImagesResult result;
+
+  if (not validateGrabImagesGoal(goal))
```
Comment (dorezyuk) [inline src/astra_driver.cpp]: please check this before calling connect
* its cheaper
* here you would return from the function without shutting down the publisher

File: src/astra_driver.cpp [hunk 26]
```diff
@@ -464,6 +579,190 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+
+  device_->setAutoExposure(goal->exposure_auto);
+
+  // TODO Check parameter units and transform parameters if necessary (orbbec
+  // doesn't provide any info on this)
+
+  actionlib_msgs::GoalStatus goal_status;
+  goal_status.status = actionlib_msgs::GoalStatus::SUCCEEDED;
+  camera_control_msgs::GrabImagesFeedback feedback;
+  size_t num_imgs =
```
Comment (dorezyuk) [inline src/astra_driver.cpp:634]: please go for the initializer list overload of std::max
```
const auto num_imsg = std::max({1, goal->exposure_times.size(), goal->gain_values.size()});
```

File: src/astra_driver.cpp [hunk 27]
```diff
@@ -464,6 +579,190 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+  // Finalize result
+  if (!result.images.empty())
+  {
+    const sensor_msgs::Image& image = result.images.front();
+    result.cam_info = *getColorCameraInfo(image.width, image.height, image.header.stamp);
+  }
+  result.success = result.images.size() == num_imgs;
+
+  // Send result
+  switch (goal_status.status)
```
Comment (dorezyuk) [inline src/astra_driver.cpp:687]: why is the setAborted case missing? I think it makes sense to do this, if we fail to get the images

File: src/astra_driver.cpp [hunk 28]
```diff
@@ -464,6 +579,190 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+  // Finalize result
+  if (!result.images.empty())
+  {
+    const sensor_msgs::Image& image = result.images.front();
+    result.cam_info = *getDepthCameraInfo(image.width, image.height, image.header.stamp);
+  }
+  result.success = result.images.size() == 1;
+
+  // Send result
+  depth_action_server_->setSucceeded(result, "OK");
```
Comment (dorezyuk) [inline src/astra_driver.cpp]: where is the failure mode?

File: src/astra_driver.cpp [hunk 29]
```diff
@@ -464,6 +579,190 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+      color_action_server_->setPreempted(result, "Preemption requested");
+      break;
+  }
+
+  setCameraParameters(old_parameters); // Restore parameters
+  imageConnectCb();                    // Disconnect (in case there are no subscribers)
+}
+
+void AstraDriver::depthGrabImagesCallback(const camera_control_msgs::GrabImagesGoalConstPtr &goal)
+{
```
Comment (dorezyuk) [inline src/astra_driver.cpp:706]: please see the comments above

File: src/astra_driver.cpp [hunk 30]
```diff
@@ -464,6 +579,190 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+    boost::mutex::scoped_lock lock(color_syncer_.mutex);
+    color_syncer_.reset(true);
+    bool got_image = color_syncer_.cv.wait_for(
+        lock, boost::chrono::seconds(1), boost::bind(&ActionImageSyncer::ready, &color_syncer_));
+    if (not got_image)
+    {
+      ROS_WARN("Failed to grab images; color stream is unresponsive");
+      break;
+    }
+    result.images.push_back(*color_syncer_.image);
```
Comment (dorezyuk) [inline src/astra_driver.cpp:664]: the problem with this code is that your image is a shared ptr. without resetting it, you will hold it also after finishing the service request

File: src/astra_driver.cpp [hunk 31]
```diff
@@ -631,28 +931,36 @@ void AstraDriver::newIRFrameCallback(sensor_msgs::ImagePtr image)
...
   {
     data_skip_color_counter_ = 0;
 
-    if (color_subscribers_)
+    if (color_subscribers_ or color_syncer_.need_image)
     {
       image->header.frame_id = color_frame_id_;
       image->header.stamp = image->header.stamp + color_time_offset_;
 
+      color_syncer_.image = image;
```
Comment (dorezyuk) [inline src/astra_driver.cpp]: you will hold an additional image in memory - no madder if you really need it or not. please add it only if you really need a image

File: src/astra_driver.cpp [hunk 32]
```diff
@@ -1110,10 +1422,12 @@ void AstraDriver::initDevice()
     try
     {
       std::string device_URI = resolveDeviceURI(device_id_);
-      #if 0
+      #if 1
       if( device_URI == "" )
```
Comment (dorezyuk) [inline src/astra_driver.cpp:1474]:
```
if(device_URI.empty()) {}
```

File: include/astra_camera/astra_driver.h [hunk 33]
```diff
@@ -69,11 +69,79 @@
...
+  }
+
+  bool ready()
+  {
+    return acquired;
+  }
+
+  boost::mutex mutex;
+  boost::condition_variable cv;
+  const int num_skip_frames_default;
```
Comment (dorezyuk) [inline include/astra_camera/astra_driver.h]: better go for unsigned values, they have a defined overflow behavior and are what you want here

File: include/astra_camera/astra_driver.h [hunk 34]
```diff
@@ -69,11 +69,79 @@
...
+  int num_skip_frames;
+  int frames_skipped;
+  bool need_image;
+  bool acquired;
+  sensor_msgs::ImagePtr image;
```
Comment (dorezyuk) [inline include/astra_camera/astra_driver.h:137]: if you reset image after consuming it you can get rid of acquired

File: include/astra_camera/astra_driver.h [hunk 35]
```diff
@@ -69,11 +69,79 @@
...
+    need_image = true;
+    acquired = false;
+  }
+
+  void check()
+  {
+    if (frames_skipped == num_skip_frames)
+    {
+      need_image = false;
+      acquired = true;
```
Comment (dorezyuk) [inline include/astra_camera/astra_driver.h:117]: why do we acquire here anything? its seems not to be related if we have the image or not

File: src/astra_driver.cpp [hunk 36]
```diff
@@ -464,6 +579,190 @@ void AstraDriver::setDepthVideoMode(const AstraVideoMode& depth_video_mode)
...
+    if (goal->gain_given)
+    {
+      device_->setIRGain(static_cast<int>(goal->gain_values[i]));
+    }
+
+    // Get image
+    boost::mutex::scoped_lock lock(color_syncer_.mutex);
+    color_syncer_.reset(true);
+    bool got_image = color_syncer_.cv.wait_for(
+        lock, boost::chrono::seconds(1), boost::bind(&ActionImageSyncer::ready, &color_syncer_));
```
Comment (dorezyuk) [inline src/astra_driver.cpp:658]: good idea to use a predicate to protect yourself against a spurios wakeup, however since your acquired is unrelated to the image, you might pass this barrier without having a valid image... the de-reference operator blow would then result in a segfault

File: test/action_image_client.cpp [hunk 37]
```diff
@@ -0,0 +1,66 @@
+#include <actionlib/client/simple_action_client.h>
+#include <camera_control_msgs/GrabImagesAction.h>
+#include <image_transport/image_transport.h>
+#include <ros/ros.h>
+
+typedef actionlib::SimpleActionClient<camera_control_msgs::GrabImagesAction>
+    ActionImageActionClient;
+typedef boost::shared_ptr<ActionImageActionClient> ActionImageActionClientPtr;
```
Comment (dorezyuk) [inline test/action_image_client.cpp]: cool that you use smart pointers, but you don't share the action server. please go for a unique_ptr so the ownership is easily trackable

**Review by dorezyuk — CHANGES_REQUESTED** [review]

File: src/astra_driver.cpp (hunk 26 above)
Comment (brinij) [reply src/astra_driver.cpp:634]: This is c++11. We are trying to get this thing merged upstream so we are changing only what we have to.

File: include/astra_camera/astra_driver.h (hunk 35 above)
Comment (brinij) [reply include/astra_camera/astra_driver.h:117]: Getting one image is not considered as an acquired state. We need to skip some frames before we go into that state. The auto-brightness of the device needs more than one image.

File: include/astra_camera/astra_driver.h (hunk 34 above)
Comment (brinij) [reply include/astra_camera/astra_driver.h:137]: See the comment below.

File: src/astra_driver.cpp (hunk 36 above)
Comment (brinij) [reply src/astra_driver.cpp:658]: One line below this one, there is a check if we got a valid image.

Comment (dorezyuk) [reply src/astra_driver.cpp:658]: nope - this is unfortunately not how it works
1. boost cv is affected by spurios wakeups : https://www.boost.org/doc/libs/1_74_0/doc/html/thread/synchronization.html
2. in a sporious wakeup the cv will trigger without a notify_all() or notify_one() call (so unrelated to the image)
3. the predicate protects you against such issues, but here its is not related to the image: consider the case, where the user sets skip_frames to false. then the predicate will return immediately true and got_image will be true

File: src/astra_driver.cpp (hunk 32 above)
Comment (brinij) [reply src/astra_driver.cpp:1474]: This would change the original code. (wish for the upstream merge)

File: include/astra_camera/astra_driver.h (hunk 35 above)
Comment (dorezyuk) [reply include/astra_camera/astra_driver.h:117]: that the issue with this code (and one of the causes of many issues below)
why don't you define a 
```
void update(const sensos_msgs::ImagePtr& _img) {
    if(need_image && frames_skipped++ == num_skip_frames){
        boost::lock_guard lk(mutex);
        image = _img;
        acquired = true;
      cv.notify_one();
    }
}
```
and use it in your image callback

File: src/astra_driver.cpp (hunk 27 above)
Comment (brinij) [reply src/astra_driver.cpp:687]: Yes, I added it.

File: include/astra_camera/astra_driver.h (hunk 35 above)
Comment (nlamprian) [reply include/astra_camera/astra_driver.h:117]: The subscriber logic interferes here. needs_image is accessed again in the image callback. So the mutex is locked once in the callback, and then the syncer is accessed.

File: src/astra_driver.cpp (hunk 36 above)
Comment (nlamprian) [reply src/astra_driver.cpp:658]: Let's first differentiate between an objection in principle and actual buggy code. Just because `acquired` and `image` do not appear in a function together, that doesn't make them unrelated. These are processed together [here](https://github.com/magazino/ros_astra_camera/blob/add_grab_image_action/src/astra_driver.cpp#L970).

File: src/astra_driver.cpp (hunk 30 above)
Comment (brinij) [reply src/astra_driver.cpp:664]: Done [here](https://github.com/magazino/ros_astra_camera/blob/add_grab_image_action/src/astra_driver.cpp#L686)

File: src/astra_driver.cpp (hunk 23 above)
Comment (dorezyuk) [reply src/astra_driver.cpp:226]: if would go for if()...else if() but up to you

File: src/astra_driver.cpp [hunk 38]
```diff
@@ -631,28 +955,43 @@ void AstraDriver::newIRFrameCallback(sensor_msgs::ImagePtr image)
 
 void AstraDriver::newColorFrameCallback(sensor_msgs::ImagePtr image)
 {
+  boost::mutex::scoped_lock lock(color_syncer_.mutex);
```
Comment (dorezyuk) [inline src/astra_driver.cpp]: looks better now. one last thing: can you move this lock in to the scope where you actually access color_mutex_.image? same for depthFrameCallback

File: src/astra_driver.cpp [hunk 39]
```diff
@@ -379,6 +491,22 @@ void AstraDriver::configCb(Config &config, uint32_t level)
   color_time_offset_ = ros::Duration(config.color_time_offset);
   depth_time_offset_ = ros::Duration(config.depth_time_offset);
 
+  // Call Stream activators whenever the always_active parameters are changed
+  if (config.always_active_rgb_stream != always_active_rgb_stream_)
+  {
+    // keep rgb stream active, without publishing,
```
Comment (nlamprian) [inline src/astra_driver.cpp]: You don't know here if the driver will need to publish. What you wanna say here is "regardless of subscribers"

**Review by dorezyuk — APPROVED** [review]
